### PR TITLE
Handle curl failure

### PR DESCRIPTION
**Pull request by andreabedini**
As we discovered, curl does not fail on http response codes >=400 by default.
This PR:
- adds `--fail` to curl's flags
- uses curl's `--write-out` to obtain a decent error message
- improves some error messages.

Sample output:

```
❯ cabal run -- foliage build
🌿 Foliage
# buildAction
Current time set to 2023-09-05T04:06:10Z. You can set a fixed time using the --current-time option.
# _sources/pkg-a/1.2/meta.toml
# curl (for fetchRemoteAsset https://httpbin.org/status/500)
foliage: Error when running Shake build system:
  at want, called at app/Foliage/CmdBuild.hs:50:7 in main:Foliage.CmdBuild
* Depends on: buildAction
  at apply1, called at app/Foliage/PrepareSource.hs:39:31 in main:Foliage.PrepareSource
* Depends on: prepareSource pkg-a-1.2 PackageVersionSpec {packageVersionTimestamp = Nothing, packageVersionSource = TarballSource {tarballSourceURI = https://httpbin.org/status/500, subdir = Nothing}, packageVersionRevisions = [], packageVersionDeprecations = [], packageVersionForce = False}
  at apply1, called at app/Foliage/RemoteAsset.hs:37:20 in main:Foliage.RemoteAsset
* Depends on: fetchRemoteAsset https://httpbin.org/status/500
  at error, called at app/Foliage/RemoteAsset.hs:108:19 in main:Foliage.RemoteAsset
* Raised the exception:
The requested URL returned error: 500
```

### CODE REVIEW

**Review by erikd — APPROVED** [review]
LGTM

**Review by angerman — APPROVED** [review]
Looks sensible.

File: app/Foliage/RemoteAsset.hs [hunk 1]
```diff
@@ -50,11 +57,67 @@ addFetchRemoteAssetRule cacheDir = addBuiltinRule noLint noIdentity run
...
+          show uri
+        ]
+  case exitCode of
+    ExitSuccess -> liftIO $ BS.readFile etagFile
+    ExitFailure c -> do
+      -- We show the curl exit code only if we cannot parse curl's write-out.
+      -- If we can parse it, we can craft a better error message.
+      case Aeson.eitherDecode out :: Either String CurlWriteOut of
+        Left err ->
+          error $
```
Comment (michaelpj) [inline app/Foliage/RemoteAsset.hs:108]: `fail`? I guess clearly something is catching `ErrorCall`, but it seems nicer to use the failure that `Action` provides.

**Review by michaelpj — APPROVED** [review]

File: app/Foliage/RemoteAsset.hs (hunk 1 above)
Comment (michaelpj) [reply app/Foliage/RemoteAsset.hs:108]: (and there's another below)

Comment (andreabedini) [reply app/Foliage/RemoteAsset.hs:108]: Totally agree with you on that.

**Comment by michaelpj** [comment]
What are we doing about the CI here? Re-evaluating doesn't seem to be having any effect :think

**Comment by angerman** [comment]
> What are we doing about the CI here? Re-evaluating doesn't seem to be having any effect :think

I've asked @cleverca22 for some help. I need to get some windows idiocity done before I can focus on something else :-(

**Pull request merged**